internal/pkg/config/config.go

```diff
@@ -284,27 +284,59 @@ func (fm funcMatcher) MatchFunction(path, receiver, name string) bool {
 	return fm.Package.MatchString(path) && fm.Receiver.MatchString(receiver) && fm.Method.MatchString(name)
 }
 
-var readFileOnce sync.Once
-var readConfigCached *Config
-var readConfigCachedErr error
+// readConfigResults reduces disk access across multiple ReadConfig calls.
+type readConfigResults struct {
+	once sync.Once
+	conf *Config
+	err  error
+}
 
-// ReadConfig reads, parses, and validates config file.
-func ReadConfig() (*Config, error) {
-	readFileOnce.Do(func() {
+// configCache safely stores readConfigResults for concurrent access.
+type configCache struct {
+	mu    sync.Mutex
+	cache map[string]*readConfigResults
+}
```

Comment:
This could probably be replaced by `sync.Map` if anyone has a strong preference.  However, from the `sync` documentation...

> The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.

> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

While we satisfy the first point in the optimization, we expect multiple reads to be occurring on the same key throughout each package's analysis passes.  As such, I've deferred to the recommendation of a plain Go map with separate lock coordination.

Reply:
No strong preference but use case (1) fits very well for us.